### PR TITLE
Teach javadoc_library about tree artifacts

### PR DESCRIPTION
**Pull request by beasleyr-vmw**
Instead of forwarding `ctx.files.srcs` to the `javadoc` command-line, we'll
take a detour through `ctx.actions.args` so that we handle (a) tree artifacts
and (b) very long argument lists.

Resolves https://github.com/google/bazel-common/issues/161.

### CODE REVIEW

File: tools/javadoc/javadoc.bzl [hunk 1]
```diff
@@ -46,6 +50,11 @@ def _javadoc_library(ctx):
         "-quiet",
     ]
 
+    # If not using `root_packages`, we'll roll up `srcs` using `args` so that
+    # we handle (a) tree artifacts and (b) very long argument lists.
+    args = ctx.actions.args()
+    args.use_param_file(param_file_arg = "@%s")
```
Comment (netdpb) [inline tools/javadoc/javadoc.bzl:56]: I think you need `use_always = True` too, right?

File: tools/javadoc/javadoc.bzl [hunk 2]
```diff
@@ -19,6 +19,10 @@ def _android_jar(android_api_level):
         return None
     return Label("@androidsdk//:platforms/android-%s/android.jar" % android_api_level)
 
+def _endswith_java(item):
```
Comment (netdpb) [inline tools/javadoc/javadoc.bzl:22]: What other files do you expect in `srcs`?

**Review by netdpb — CHANGES_REQUESTED** [review]

File: tools/javadoc/javadoc.bzl [hunk 3]
```diff
@@ -46,6 +50,11 @@ def _javadoc_library(ctx):
         "-quiet",
     ]
 
+    # If not using `root_packages`, we'll roll up `srcs` using `args` so that
+    # we handle (a) tree artifacts and (b) very long argument lists.
+    args = ctx.actions.args()
```
[the remark this answers is not in the document]
Comment (beasleyr-vmw) [reply tools/javadoc/javadoc.bzl]: Sounds good. Will do.

File: tools/javadoc/javadoc.bzl (hunk 1 above)
Comment (beasleyr-vmw) [reply tools/javadoc/javadoc.bzl]: As I understand it, it depends. Without `use_always`, Bazel will expand short arg lists on the command-line and use the args file only for long arg lists. I don't know enough about Blaze/Bazel internals to have a strong opinion, but I have the impression that Google prefers to generate parameter files only when necessary.

It's a moot point w/ the latest revision, regardless. :)

File: tools/javadoc/javadoc.bzl (hunk 2 above)
Comment (beasleyr-vmw) [reply tools/javadoc/javadoc.bzl:22]: In my case, my code generator produces both `.java` and `.xml`. One could imagine another tree artifact consisting of Java sources from `.../src/main/java/**/*.java` as well as any misc. resources from `.../src/main/resources/**`.

The `javadoc` utility will choke if any input file doesn't have a `.java` extension, so either I could (within reason) allow for some flexibility in this rule, or I could push the responsibility of filtering onto the users. Since it's cheap/easy, I opted w/ the former.

File: tools/javadoc/javadoc.bzl [hunk 4]
```diff
@@ -62,45 +61,76 @@ def _javadoc_library(ctx):
...
-        outputs = [output_dir, ctx.outputs.jar],
-        arguments = [args],
+        outputs = [output_dir],
+        arguments = [
+            output_dir.path,
+            "true" if ctx.attr.root_packages else "false",
+            jargs,
+        ],
+        command = """
+outdir=$1; shift
```
Comment (netdpb) [inline tools/javadoc/javadoc.bzl:99]: I don't see where this is used.

File: tools/javadoc/javadoc.bzl [hunk 5]
```diff
@@ -62,45 +61,76 @@ def _javadoc_library(ctx):
...
+root_packages=$1; shift
+jargsfile=$1; shift
+
+# See also `TODO(b/167433657)` in javadoc.bzl.
+if $root_packages; then
```
Comment (netdpb) [inline tools/javadoc/javadoc.bzl:104]: Hmm. It would be great not to have so much shell in here. What about this?

```
command = "${JAVA_HOME}/bin/javadoc $1" +
    (' -sourcepath "$(find * -type d -name '*java' -print0 | tr '\\0' :)"' if ctx.attr.root_packages else ""),
args = [args]
```

File: tools/javadoc/javadoc.bzl [hunk 6]
```diff
@@ -62,45 +61,76 @@ def _javadoc_library(ctx):
...
+            output_dir.path,
+            "true" if ctx.attr.root_packages else "false",
+            jargs,
+        ],
+        command = """
+outdir=$1; shift
+root_packages=$1; shift
+jargsfile=$1; shift
+
+# See also `TODO(b/167433657)` in javadoc.bzl.
```
Comment (netdpb) [inline tools/javadoc/javadoc.bzl:103]: Maybe move the comment from above down to here instead?

File: tools/javadoc/javadoc.bzl [hunk 7]
```diff
@@ -62,45 +61,76 @@ def _javadoc_library(ctx):
...
+fi
+
+${JAVA_HOME}/bin/javadoc $jargsfile ${sourcepath_args[@]}
+""",
+        env = {
+            "JAVA_HOME": java_home,
+        },
+    )
+
+    # Zip the outputs using Bazel's deterministic `zipper` utility.
```
Comment (netdpb) [inline tools/javadoc/javadoc.bzl:118]: Can we do this in a separate PR, just to keep this PR to a single purpose? With the old `jar` command, the shell script bit was much simpler because it supports `-C`, which is nice.

If a shell command gets to more than a couple of lines, I'd prefer to extract it out to a separate file. I think this may not be worth it, but we can talk about the value of using `zipper` separately.

**Review by netdpb — CHANGES_REQUESTED** [review]
Thanks!

File: tools/javadoc/javadoc.bzl (hunk 1 above)
Comment (netdpb) [reply tools/javadoc/javadoc.bzl]: I see.

File: tools/javadoc/javadoc.bzl [hunk 8]
```diff
@@ -38,22 +42,17 @@ def _javadoc_library(ctx):
...
-    # we handle (a) tree artifacts and (b) very long argument lists.
-    args = ctx.actions.args()
-    args.use_param_file(param_file_arg = "@%s")
+    # javadoc args
+    # - Should we use `ctx.host_configuration.host_path_separator` instead of `:`?
+    jargs = ctx.actions.args()
+    jargs.use_param_file(param_file_arg = "@%s", use_always = True)
+    jargs.add("-use")
+    jargs.add("-encoding", "UTF8")
+    jargs.add_joined("-classpath", [jar.path for jar in classpath], join_with = ":")
```
[the remark this answers is not in the document]
Comment (netdpb) [reply tools/javadoc/javadoc.bzl:51]: Right.

File: tools/javadoc/javadoc.bzl [hunk 9]
```diff
@@ -61,76 +58,53 @@ def _javadoc_library(ctx):
...
-fi
-
-${JAVA_HOME}/bin/javadoc $jargsfile ${sourcepath_args[@]}
-""",
-        env = {
-            "JAVA_HOME": java_home,
-        },
+        arguments = [args],
+        command = "${JAVA_HOME}/bin/javadoc $1" + (
+            # See TODO(b/167433657) above.
```
Comment (netdpb) [inline tools/javadoc/javadoc.bzl:91]: Let's just move the TODO and the comment below it to here, since it all applies to this bit.

File: tools/javadoc/javadoc.bzl [hunk 10]
```diff
@@ -61,76 +58,53 @@ def _javadoc_library(ctx):
...
+            ' -sourcepath "$(find * -type d -name \'*java\' -print0 | tr \'\\0\' :)"' if ctx.attr.root_packages else ""
```
Comment (netdpb) [inline tools/javadoc/javadoc.bzl:92]:
```suggestion
            """ -sourcepath "$(find * -type d -name '*java' -print0 | tr '\\0' :)"""" if ctx.attr.root_packages else ""
```

Does this work? A little less backslash-escaping.

File: tools/javadoc/tests.bzl [hunk 11]
```diff
@@ -97,13 +186,17 @@ def _root_packages_test_case(name):
...
 def analysis_test_suite(name):
+    _deps_test_case(name = name + "_dep")
     _group_test_case(name = name + "_group")
+    _links_test_case(name = name + "_links")
     _root_packages_test_case(name = name + "_root_packages")
 
     native.test_suite(
         name = name,
         tests = [
+            ":%s_dep_test" % name,
```
Comment (netdpb) [inline tools/javadoc/tests.bzl:197]: You don't need the colon, so you could create variables you can reuse here and when you define the test case.

**Review by netdpb — COMMENTED** [review]
Looking good! Thanks!

File: tools/javadoc/tests.bzl (hunk 11 above)
Comment (beasleyr-vmw) [reply tools/javadoc/tests.bzl:197]: I don't understand. In any event, I don't _have_ to parameterize by `name` (it's unlikely this package will have multiple suites where uniqueness matters), so I'll just keep it simple and stick to literals like `deps_test` and `group_test`.
